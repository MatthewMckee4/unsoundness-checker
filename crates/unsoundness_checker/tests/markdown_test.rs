use std::{fs, panic, path::Path};

pub mod common;

use common::run_rule_tests;

#[test]
fn test_all_rules_from_markdown() {
    let rules_dir = "resources/rules";

    let entries = fs::read_dir(rules_dir).expect("Failed to read rules directory");
    let mut rule_files = Vec::new();

    for entry in entries {
        let entry = entry.expect("Failed to read directory entry");
        let path = entry.path();
        if path.extension().and_then(|s| s.to_str()) == Some("md")
            && let Some(rule_name) = path.file_stem().and_then(|s| s.to_str())
        {
            rule_files.push(rule_name.to_string());
        }
    }

    assert!(
        !rule_files.is_empty(),
        "No rule markdown files found in {rules_dir}",
    );

    rule_files.sort();

    let mut failures = Vec::new();

    for rule_name in rule_files {
        let results = run_rule_tests(&rule_name);

        let snapshots_dir = format!("tests/snapshots/{rule_name}");

        fs::create_dir_all(&snapshots_dir)
            .unwrap_or_else(|_| panic!("Failed to create snapshots directory for {rule_name}"));

        for (temp_path, snippet_name, output) in results {
            let temp_filter = tempdir_filter(&temp_path);

            eprint!("test {rule_name}/{snippet_name}");

            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path(format!("snapshots/{rule_name}"));
            settings.add_filter(&temp_filter, "<temp_dir>/");

            let result = panic::catch_unwind(panic::AssertUnwindSafe(|| {
                settings.bind(|| {
                    insta::assert_snapshot!(snippet_name.clone(), output);
                });
            }));

            if let Err(err) = result {
                let error_message = err.downcast_ref::<String>().map_or_else(
                    || {
                        err.downcast_ref::<&str>()
                            .map_or_else(|| "Unknown panic".to_string(), |s| (*s).to_string())
                    },
                    Clone::clone,
                );
                failures.push(format!("{rule_name}/{snippet_name}: {error_message}"));
                eprintln!(" failed");
            } else {
                eprintln!(" passed");
            }
        }
    }

    assert!(
        failures.is_empty(),
        "Snapshot assertions failed:\n{}",
        failures.join("\n")
    );
}

fn tempdir_filter(path: &Path) -> String {
    format!(r"{}\\?/?", regex::escape(path.to_str().unwrap()))
}
