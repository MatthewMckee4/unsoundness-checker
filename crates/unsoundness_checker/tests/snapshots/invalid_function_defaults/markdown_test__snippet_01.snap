---
source: crates/unsoundness_checker/tests/markdown_test.rs
expression: output
---
error[invalid-function-defaults]: Setting `__defaults__` to an object of type `tuple[Literal["string"]]` on a function may lead to runtime type errors.
 --> snippet_01.py:4:1
  |
2 |     return x
3 |
4 | foo.__defaults__ = ("string",)
  | ^^^^^^^^^^^^^^^^
5 |
6 | foo.__defaults__ = None
  |
info: rule `invalid-function-defaults` was selected in the configuration file

error[invalid-function-defaults]: Setting `__defaults__` to an object of type `None` on a function may lead to runtime type errors.
 --> snippet_01.py:6:1
  |
4 | foo.__defaults__ = ("string",)
5 |
6 | foo.__defaults__ = None
  | ^^^^^^^^^^^^^^^^
7 |
8 | def bar(x: int = 1, y: str = "string", z: bool = True) -> int:
  |
info: rule `invalid-function-defaults` was selected in the configuration file

error[invalid-function-defaults]: Setting `__defaults__` to an object of type `tuple[Literal[1], Literal["string"], Literal["string"]]` on a function may lead to runtime type errors.
  --> snippet_01.py:11:1
   |
 9 |     return x
10 |
11 | bar.__defaults__ = (1, "string", "string")
   | ^^^^^^^^^^^^^^^^
12 |
13 | bar.__defaults__ = None
   |
info: rule `invalid-function-defaults` was selected in the configuration file

error[invalid-function-defaults]: Setting `__defaults__` to an object of type `None` on a function may lead to runtime type errors.
  --> snippet_01.py:13:1
   |
11 | bar.__defaults__ = (1, "string", "string")
12 |
13 | bar.__defaults__ = None
   | ^^^^^^^^^^^^^^^^
14 |
15 | bar.__defaults__ = (1, "string")
   |
info: rule `invalid-function-defaults` was selected in the configuration file

error[invalid-function-defaults]: Setting `__defaults__` to an object of type `tuple[Literal[1], Literal["string"]]` on a function may lead to runtime type errors.
  --> snippet_01.py:15:1
   |
13 | bar.__defaults__ = None
14 |
15 | bar.__defaults__ = (1, "string")
   | ^^^^^^^^^^^^^^^^
   |
info: rule `invalid-function-defaults` was selected in the configuration file
